proposals/0000-lambda-layout.md

```diff
@@ -0,0 +1,384 @@
+    | result == GL_REPEAT -> Repeat
+    | otherwise -> error "getWrapping: unexpected wrapping mode."
+```
+
+red-black-record-2.1.0.3/lib/Data/RBR/Internal.hs
+```Haskell
+_prefixNS = \case
+    Left  l -> S l
+    Right x -> case x of Here fv -> Z @_ @v @start fv
+_breakNS = \case
+    Z x -> Right (Here x)
+    S x -> Left x
+
+-- with -XMultiWayLambda
+_prefixNS = \of
+    (Left  l) -> S l
+    (Right x) -> case x of Here fv -> Z @_ @v @start fv
+_breakNS = \of
+    (Z x) -> Right (Here x)
+    (S x) -> Left x
+```
+
+recursors-0.1.0.0/Control/Final.hs
+```Haskell
+map (\case PlainTV n    -> n
+           KindedTV n _ -> n) binders
+           
+-- With -XMultiWayLambda
+
+map (\of (PlainTV n)    -> n
+         (KindedTV n _) -> n) binders
+```
+
+roc-id-0.1.0.0/library/ROC/ID/Gender.hs
+```Haskell
+printGender :: Language -> Gender -> Text
+printGender = \case
+  English -> printGenderEnglish
+  Chinese -> printGenderChinese
+
+printGenderEnglish :: Gender -> Text
+printGenderEnglish = \case
+  Male   -> "Male"
+  Female -> "Female"
+
+printGenderChinese :: Gender -> Text
+printGenderChinese = \case
+  Male   -> "男性"
+  Female -> "女性"
+
+-- With -XMultiWayLambda - this makes use of the capability to have multiple parameters
+
+printGender :: Language -> Gender -> Text
+printGender = \of
+  English Male   -> "Male"
+  English Female -> "Female"
+  Chinese Male   -> "男性"
+  Chinese Female -> "女性"
+```
+
+## Effect and Interactions
+
+Enabling the extension enables users to use the suggested syntax. This obviates
+the need for `-XMultiWayIf` and `-XLambdaCase`.
+
+As `of` is already a keyword, no currently allowed syntax is stolen by this extension,
+and the behavior of no currently legal program would be changed with the extension
+enabled.
+
+## Costs and Drawbacks
+
+It is one additional syntactic construct to maintain, however the maintenance
+cost should be fairly low due to the similarity to already existing constructs.
+
+While this also means one additional construct to learn for beginners, the
+syntax is consistent with similar constructs in the existing language, and as
+such users might be surprised that a construct with these capabilities
+doesn't yet exist.
+
+## Alternatives
+
+ - Zero clauses could be permitted. In this case, however, a way would have to be found
+   to indicate how many arguments a given `\ of`-expression matches on, as otherwise, it would
+   be ambiguous.
+   The number of arguments a `\ of`-expression pattern matches on becomes obvious from the
+   clauses, e.g. `\ of a b -> ...` clearly matches on two arguments. Without clauses, this remains
+   unclear. This means it would also be unclear whether the patterns are non-exhaustive:
+   Consider the expression `\of {} :: Bool -> Void -> a`. If the expression is supposed to match on
+   both arguments, the patterns are exhaustive. If it is only supposed to match on the first argument
+   and evaluate to a funtion of type `Void -> a`, it is not exhaustive.
```

Comment:
Besides exhaustiveness, the number of arguments also changes how many arguments it must be applied to before bottoming out:

Is ``(\of {}) `seq` ()`` `()` or `undefined`? How about `(\of {}) () `seq` ()`?

(It is not an entirely different observation, but from the point of view of the dynamic semantics, not the static semantics)

Reply:
Hmm actually I think a empty-case `\of` is never "more bottom" than any of it's arguments. If the must-be-bottom argument has yet to be applied, the expression should be non-bottom.

On the other hand, it's unclear whether or not a multi-argumemt empty case forces it's other arguments, or which bottom-only argument is forced if there are multiple.

These ambiguities also apply to regular function binds making them no less awkward for empty case, so I am not concerned. We'll find a solution for both some day.

Reply:
It will be “more bottom” than its arguments if its not exhaustive.

But even if the `\of` is only applied to arguments of type `Void` or equivalent (hence exhaustive), you have semantic differences. At type `Void -> Void`, `undefined` is an observably different point than `\x -> x` (the latter being the same as `\case {}`).

Reply:
> It will be “more bottom” than its arguments if its not exhaustive.

Ah true, I did not think of incomplete 0 alternative pattern matching.

> But even if the `\of` is only applied to arguments of type `Void` or equivalent (hence exhaustive), you have semantic differences. At type `Void -> Void`, `undefined` is an observably different point than `\x -> x` (the latter being the same as \case {}).

Ah I think I know what you meant all along. `\of {}`, unapplied to any arguments, should never be bottom unless it's the 0 scrutenees, 0 alternatives, incomplete pattern match.